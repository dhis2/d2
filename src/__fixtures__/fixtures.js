export default (function x() {
    const fixtures = {};

    function getFixture(fixtureName) {
        if (fixtures && fixtures[fixtureName]) {
            // Make sure we return a new object
            return JSON.parse(JSON.stringify(fixtures[fixtureName]));
        }
        throw new Error(['Fixture', fixtureName, 'does not exist'].join(' '));
    }

    function addFixture(fixtureName, fixtureData) {
        if (fixtureName && fixtureData && typeof fixtureName === 'string') {
            fixtures[fixtureName] = fixtureData;
        }
    }

    addFixture('/api/attributes', require('./json/api/schemas/attributes.json'));
    addFixture('/api/dataElements', require('./json/api/dataElements.json'));
    addFixture('/api/organisationUnits', require('./json/api/organisationUnits.json'));
    addFixture('/api/schemas', require('./json/api/schemas.json'));
    addFixture('/api/schemas/dataElement', require('./json/api/schemas/dataElement.json'));
    addFixture('/api/schemas/dataSet', require('./json/api/schemas/dataSet'));
    addFixture('/api/schemas/indicatorGroup', require('./json/api/schemas/indicatorGroup.json'));
    addFixture('/api/schemas/organisationUnit', require('./json/api/schemas/organisationUnit.json'));
    addFixture('/api/schemas/organisationUnitGroupSet', require('./json/api/schemas/organisationUnitGroupSet'));
    addFixture('/api/schemas/validationRule', require('./json/api/schemas/validationRule'));
    addFixture('/api/schemas/program', require('./json/api/schemas/program'));
    addFixture('/api/schemas/programNotificationTemplate', require('./json/api/schemas/programNotificationTemplate'));
    addFixture('/api/schemas/user', require('./json/api/schemas/user.json'));
    addFixture('/api/schemas/legendSet', require('./json/api/schemas/legendSet.json'));
    addFixture('/api/schemas/legend', require('./json/api/schemas/legend.json'));
    addFixture('/api/userSettings', require('./json/api/userSettings'));
    addFixture('/appStore', require('./json/api/appStore.json'));
    addFixture('/dataElementAttributes', require('./json/dataElementAttributes.json'));
    addFixture('/me', require('./json/api/me.json'));
    addFixture('/me/authorities', require('./json/api/me/authorities.json'));
    addFixture('/modelDefinitions/dataElement', require('./json/modelDefinitions/dataElement.json'));
    addFixture('/singleUserAllFields', require('./json/singleUserAllFields.json'));
    addFixture('/singleUserOwnerFields', require('./json/singleUserOwnerFields.json'));
    addFixture('/api/legendSets/k1JHPfXsJND', require('./json/api/legendSets/k1JHPfXsJND.json'));
    addFixture('/api/analytics/dataValueSet', require('./json/api/analytics/dataValueSet.json'));
    addFixture('/api/analytics/rawData', require('./json/api/analytics/rawData.json'));
    addFixture('/api/analytics/aggregate', require('./json/api/analytics/aggregate.json'));
    addFixture('/api/analytics/count', require('./json/api/analytics/count.json'));
    addFixture('/api/analytics/query', require('./json/api/analytics/query.json'));
    addFixture('/api/analytics/cluster', require('./json/api/analytics/cluster.json'));

    return {
        get: getFixture,
        add: addFixture,
    };
}());
