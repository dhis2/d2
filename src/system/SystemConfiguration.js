import Api from '../api/Api'

/**
 * Handles communication with the configuration endpoint. Can be used to get or set configuration options.
 *
 * @memberof module:system
 */
class SystemConfiguration {
    constructor(api = Api.getApi()) {
        this.api = api

        this.configuration = undefined
        this.configPromise = null
    }

    /**
     * Fetches all system configuration settings from the API and caches them so that future
     * calls to this function won't call the API again.
     *
     * @param {boolean} [ignoreCache=false] If set to true, calls the API regardless of cache status
     * @returns {Promise} Promise that resolves with all the individual configuration options from the api.
     */
    all(ignoreCache) {
        if (this.configPromise === null || ignoreCache === true) {
            this.configPromise = this.api
                .get('configuration')
                .then(configuration => {
                    this.configuration = configuration
                    return this.configuration
                })
        }

        return this.configPromise
    }

    /**
     * Returns the value of the specified configuration option.
     *
     * This is a convenience method that works exactly the same as calling `configuration.all()[name]`.
     *
     * @param key {String}
     * @param {boolean} [ignoreCache=false] If set to true, calls the API regardless of cache status
     * @returns {Promise}
     */
    get(key, ignoreCache) {
        return this.all(ignoreCache).then(config => {
            if (config.hasOwnProperty(key)) {
                return Promise.resolve(config[key])
            }

            return Promise.reject(`Unknown config option: ${key}`)
        })
    }

    /**
     * Send a query to the API to change the value of a configuration key to the specified value.
     *
     * @param key {String}
     * @param value {String|null}
     * @returns {Promise}
     */
    /* eslint-disable complexity */
    set(key, value) {
        const that = this
        let req

        if (key === 'systemId') {
            return Promise.reject("The system ID can't be changed")
        }
        if (
            (key === 'feedbackRecipients' ||
                key === 'selfRegistrationOrgUnit' ||
                key === 'selfRegistrationRole') &&
            (value === 'null' || value === null)
        ) {
            // Only valid UIDs are accepted when POST'ing, so we have to use DELETE in stead of POST'ing a null value.
            req = this.api.delete(['configuration', key].join('/'))
        } else if (key === 'corsWhitelist') {
            // The corsWhitelist endpoint expects a JSON array (of URLs), while here value is expected to be a string.
            req = this.api.post(
                ['configuration', key].join('/'),
                value.trim().split('\n')
            )
        } else {
            req = this.api.post(['configuration', key].join('/'), value, {
                dataType: 'text',
                contentType: 'text/plain',
            })
        }

        return req
            .then(() => {
                // Ideally we'd update the cache here, but doing so requires another trip to the server
                // For now, just bust the cache to ensure it's not incorrect
                that.configuration = undefined
                return Promise.resolve()
            })
            .catch(() => Promise.reject(`No configuration found for ${key}`))
    }
    /* eslint-enable complexity */
}

export default SystemConfiguration
