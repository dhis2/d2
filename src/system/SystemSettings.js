/**
 * @module System
 *
 * @requires lib/check
 * @requires api/Api
 */

import { isString } from '../lib/check';
import Api from '../api/Api';


/**
 * @class SystemSettings
 *
 * @description
 * Handles communication with the systemSettings endpoint. Can be used to get or save systemSettings.
 */
// TODO: Return the values from the local cache if we have not updated it? We could
class SystemSettings {
    constructor(api = Api.getApi()) {
        this.api = api;
    }

    /**
     * @method all
     *
     * @returns {Promise} Promise that resolves with the systemsettings object from the api.
     *
     * @description
     * Loads all the system settings in the system and returns them as an object from the promise.
     * ```js
     * d2.system.settings.all()
     *  .then(systemSettings => {
     *    console.log('Analytics was last updated on: ' + systemSettings.keyLastSuccessfulResourceTablesUpdate);
     *  });
     * ```
     */
    all() {
        return this.settings
            ? Promise.resolve(this.settings)
            : this.api.get('systemSettings')
                .then((settings) => { this.settings = settings; return Promise.resolve(this.settings); });
    }

    /**
     * @method get
     *
     * @param {String} systemSettingsKey The identifier of the system setting that should be retrieved.
     * @returns {Promise} A promise that resolves with the value or will fail if the value is not available.
     *
     * @description
     * ```js
     * d2.system.settings.get('keyLastSuccessfulResourceTablesUpdate')
     *  .then(systemSettingsValue => {
     *    console.log('Analytics was last updated on: ' + systemSettingsValue);
     *  });
     * ```
     */
    get(systemSettingsKey) {
        if (this.settings && this.settings[systemSettingsKey]) {
            return Promise.resolve(this.settings[systemSettingsKey]);
        }

        function processValue(value) {
            // Attempt to parse the response as JSON. If this fails we return the value as is.
            try {
                return JSON.parse(value);
            } catch (e) {
                return value;
            }
        }

        return new Promise((resolve, reject) => {
            if (!isString(systemSettingsKey)) {
                throw new TypeError('A "key" parameter should be specified when calling get() on systemSettings');
            }

            const options = { headers: { accept: 'text/plain' } };
            this.api.get(
                ['systemSettings', systemSettingsKey].join('/'), undefined, options)
                .then((response) => {
                    if (response) {
                        resolve(processValue(response));
                    }
                    reject(new Error('The requested systemSetting has no value or does not exist.'));
                });
        });
    }

    set(systemSettingsKey, value) {
        delete this.settings;

        const settingUrl = ['systemSettings', systemSettingsKey].join('/');
        if (value === null || (`${value}`).length === 0) {
            return this.api.delete(settingUrl);
        }
        return this.api.post(settingUrl, value, { headers: { 'Content-Type': 'text/plain' } });
    }
}

export default SystemSettings;
