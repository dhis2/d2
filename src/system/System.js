/**
 * @module system
 *
 * @requires module:api/Api
 */

import Api from '../api/Api';
import SystemSettings from './SystemSettings';
import SystemConfiguration from './SystemConfiguration';

/**
 * @description
 * Represents the system that can be interacted with. There is a single instance of this pre-defined onto the d2
 * object after initialisation. This can be interacted with using its property objects to among other be used
 * to get and save systemSettings.
 *
 * @memberof module:system
 */
class System {
    constructor(settings, configuration) {
        /**
         * Contains a reference to a `SystemSettings` instance that can be used
         * to retrieve and save system settings.
         *
         * ```js
         * d2.system.settings.get('keyLastSuccessfulResourceTablesUpdate')
         *  .then(systemSettingsValue => {
         *    console.log('Analytics was last updated on: ' + systemSettingsValue);
         *  });
         * ```
         * @type {SystemSettings}
         *
         */
        this.settings = settings;

        /**
         * A representation of the system configuration,
         * that can be used to retrieve and change system configuration options.
         * @type {SystemConfiguration}
         */
        this.configuration = configuration;

        /**
         * An object containing system information about the DHIS2 instance
         * @type {Object}
         */
        this.systemInfo = undefined;

        /**
         * An object containing version information about the DHIS2 instance
         * @type {Object}
         */
        this.version = undefined;

        /**
         * An array of all the webapps that are installed on the current DHIS2 instance
         * @type {Array}
         */
        this.installedApps = undefined;
    }

    /**
     * Sets the systemInfo and version properties
     *
     * @param systemInfo
     */
    setSystemInfo(systemInfo) {
        this.version = System.parseVersionString(systemInfo.version);
        this.systemInfo = systemInfo;
    }

    /**
     * Sets the list of currently installed webapps
     *
     * @param apps
     */
    setInstalledApps(apps) {
        this.installedApps = apps;
    }

    /**
     * Refreshes the list of currently installed webapps
     *
     * @returns {Promise} A promise that resolves to the list of installed apps
     */
    loadInstalledApps() {
        const api = Api.getApi();

        return api.get('apps')
            .then((apps) => {
                this.setInstalledApps(apps);

                return apps;
            });
    }

    /**
     * Upload and install a zip file containing a new webapp
     *
     * @param zipFile Zip file data from a file input form field
     * @param onProgress An optional callback that will be called whenever file upload progress info is available
     * @returns {Promise}
     */
    uploadApp(zipFile, onProgress) { // eslint-disable-line class-methods-use-this
        const api = Api.getApi();
        const data = new FormData();
        let xhr;
        data.append('file', zipFile);

        if (onProgress !== undefined) {
            xhr = new XMLHttpRequest();
            xhr.upload.onprogress = (progress) => {
                if (progress.lengthComputable) {
                    onProgress(progress.loaded / progress.total);
                }
            };
        }

        return api.post('apps', data, {
            contentType: false,
            processData: false,
            xhr: xhr !== undefined ? () => xhr : undefined,
        });
    }

    /**
     * Load the list of apps available in the DHIS 2 app store
     *
     * @param compatibleOnly If true, apps that are incompatible with the current system will be filtered out
     * @returns {Promise}
     */
    loadAppStore(compatibleOnly = true) {
        return new Promise((resolve, reject) => {
            const api = Api.getApi();
            api.get('appStore')
                .then(appStoreData => resolve(appStoreData
                    .map((appData) => {
                        const app = Object.assign({}, appData);

                        if (compatibleOnly) {
                            app.versions = app.versions
                                .filter(versionData => System.isVersionCompatible(this.version, versionData));
                        }

                        return app;
                    })
                    .filter(appData => appData.versions.length > 0)),
                )
                .catch(err => reject(err));
        });
    }

    /**
     * Install the specified app version from the DHIS 2 app store
     *
     * @param uid The uid of the app version to install
     * @returns {Promise}
     */
    installAppVersion(uid) { // eslint-disable-line class-methods-use-this
        const api = Api.getApi();
        return new Promise((resolve, reject) => {
            api.post(['appStore', uid].join('/'), '', { dataType: 'text' }).then(() => {
                resolve();
            }).catch((err) => {
                reject(err);
            });
        });
    }

    /**
     * Remove the specified app from the system
     *
     * @param appKey The key of the app to remove
     * @returns {Promise}
     */
    uninstallApp(appKey) { // eslint-disable-line class-methods-use-this
        const api = Api.getApi();

        return api.delete(['apps', appKey].join('/'))
            // TODO: Stop jQuery from rejecting successful promises
            .catch(() => undefined);
    }

    /**
     * Refresh the list of apps that are installed on the server
     *
     * @returns {Promise} A promise that resolves to the updated list of installed apps
     */
    reloadApps() {
        const api = Api.getApi();
        return api.update('apps').then(() => this.loadInstalledApps());
    }

    /**
     * @static
     * @typedef {Object} ParsedVersion
     * @property {number} major - Major version of the parsed-string (before .)
     * @property {number} minor - Minor version of the parsed-string (after .)
     * @property {boolean} snapshot - If it's a snapshot-version
     */
    // TODO: Validate string
    // TODO: Handle valid version objects too
    /**
     * Parses a version string into an object describing the version.
     * @param {string} version Version-string to parse
     * @returns {module:system.ParsedVersion}
     */
    static parseVersionString(version) {
        return {
            major: Number.parseInt(version, 10),
            minor: Number.parseInt(version.substring(version.indexOf('.') + 1), 10),
            snapshot: version.indexOf('-SNAPSHOT') >= 0,
        };
    }

    // Disable eslint complexity warning
    /* eslint-disable complexity */
    /**
     * Compares version a to version b.
     * @param a {string|module:system.ParsedVersion}
     * @param b {string|module:system.ParsedVersion}
     * @returns {number} 0 if same version, else a - b.
     */
    static compareVersions(a, b) {
        const from = (typeof a === 'string' || a instanceof String) ? System.parseVersionString(a) : a;
        const to = (typeof b === 'string' || b instanceof String) ? System.parseVersionString(b) : b;

        if (from.major !== to.major) {
            return from.major - to.major;
        } else if (from.minor !== to.minor) {
            return from.minor - to.minor;
        }

        return (from.snapshot ? 0 : 1) - (to.snapshot ? 0 : 1);
    }

    /**
     * Checks if systemVersion is compatible with appVersion.
     * Versions are compatible if appVersion.minDhisVersion <= parsed systemVersion and
     * if appVersion.maxDhisVersion >= parsed systemVersion.
     * @param systemVersion {string|module:system.ParsedVersion} systemVersion to check
     * @param {Object} appVersion - AppVersion object to check
     * @returns {boolean} true if compatible, false otherwise.
     */
    static isVersionCompatible(systemVersion, appVersion) {
        const minVersion = appVersion.minDhisVersion || appVersion.min_platform_version || null;
        const maxVersion = appVersion.maxDhisVersion || appVersion.max_platform_version || null;

        const isNewEnough = (minVersion ? System.compareVersions(systemVersion, minVersion) >= 0 : true);
        const isNotTooOld = (maxVersion ? System.compareVersions(systemVersion, maxVersion) <= 0 : true);

        return isNewEnough && isNotTooOld;
    }
    /* eslint-enable */

    /**
     * Get a new instance of the system object. This will function as a singleton, when a System object has been created
     * when requesting getSystem again the original version will be returned.
     *
     * @returns {System} Object with the system interaction properties
     */
    static getSystem() {
        if (!System.getSystem.system) {
            System.getSystem.system = new System(new SystemSettings(), new SystemConfiguration());
        }

        return System.getSystem.system;
    }
}

export default System;
