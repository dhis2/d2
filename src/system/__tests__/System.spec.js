import fixtures from '../../__fixtures__/fixtures';
import MockApi from '../../api/Api';
import System from '../System';
import SystemConfiguration from '../SystemConfiguration';
import SystemSettings from '../SystemSettings';

jest.mock('../../api/Api');

describe('System', () => {
    let system;
    let apiMock;

    beforeEach(() => {
        apiMock = new MockApi();
        system = new System(new SystemSettings(), new SystemConfiguration());
    });

    afterEach(() => {
        MockApi.mockReset();
    });

    it('should be an instance of System', () => {
        expect(system).toBeInstanceOf(System);
    });

    it('should not be allowed to be called without new', () => {
        expect(() => System()).toThrowErrorMatchingSnapshot();
    });

    it('should contain an instance of SystemConfiguration', () => {
        expect(system.configuration).toBeInstanceOf(SystemConfiguration);
    });

    it('should contain an instance of SystemSettings', () => {
        expect(system.settings).toBeInstanceOf(SystemSettings);
    });

    describe('loadInstalledApps()', () => {
        let appsFromApi;

        beforeEach(() => {
            appsFromApi = [
                {
                    version: '0.4.8',
                    name: 'Data Approval',
                    description: 'Approvals app for PEPFAR',
                    icons: {
                        48: 'img/icons/dataapproval.png',
                    },
                    developer: {
                        url: 'http://www.dhis2.org',
                        name: 'Mark Polak',
                        company: 'DHIS2 Core Team',
                        email: 'markpo@ifi.uio.no',
                    },
                    activities: {
                        dhis: {
                            href: 'http://localhost:8080/dhis',
                        },
                    },
                    folderName: 'approvals',
                    launchUrl: 'http://localhost:8080/dhis/api/apps/approvals/index.html?v=0.4.8',
                    key: 'approvals',
                    launch_path: 'index.html?v=0.4.8',
                    default_locale: 'en',
                },
                {
                    version: '0.0.1',
                    name: 'Data Export Log',
                    description: 'Data export log viewer',
                    icons: {
                        48: 'icons/export.png',
                    },
                    developer: {
                        url: '',
                        name: 'Mark Polak',
                    },
                    activities: {
                        dhis: {
                            href: 'http://localhost:8080/dhis',
                        },
                    },
                    folderName: 'data-export-log',
                    launchUrl: 'http://localhost:8080/dhis/api/apps/data-export-log/index.html?1.0.0-rc1',
                    key: 'data-export-log',
                    launch_path: 'index.html?1.0.0-rc1',
                    default_locale: 'en',
                },
            ];

            apiMock.get = jest.fn().mockReturnValue(Promise.resolve(appsFromApi));
        });

        it('should set the list of installed apps onto the Settings object', () => system.loadInstalledApps()
            .then(() => {
                expect(system.installedApps).toEqual(appsFromApi);
            }));

        it('should reject the promise if the request fails', () => {
            apiMock.get = jest.fn().mockReturnValue(Promise.reject('Apps can not be loaded'));

            return system.loadInstalledApps()
                .catch(error => error)
                .then((message) => {
                    expect(message).toBe('Apps can not be loaded');
                });
        });

        it('should resolve with the returned list of apps', () => system.loadInstalledApps()
            .then((apps) => {
                expect(apps).toEqual(appsFromApi);
            }));
    });

    describe('uploadApp()', () => {
        let appendSpy;
        let formData;

        beforeEach(() => {
            jest.spyOn(apiMock, 'post')
                .mockReturnValue(Promise.resolve());

            // Fake FormData object
            appendSpy = jest.fn();
            formData = {
                append: appendSpy,
            };
            // Fake formData global constructor
            global.FormData = function FormData() {
                return formData;
            };
        });

        afterEach(() => {
            global.FormData = undefined;
        });

        it('should be a function on the system object', () => {
            expect(typeof system.uploadApp).toBe('function');
        });

        it('should call the post with the correct options', () => {
            const xhrOptions = {
                contentType: false,
                processData: false,
                xhr: undefined,
            };

            system.uploadApp('ZipFile');

            expect(apiMock.post).toBeCalledWith('apps', formData, xhrOptions);
        });

        it('should call append on the formData object to add the file to upload', () => {
            system.uploadApp('ZipFile');

            expect(formData.append).toBeCalledWith('file', 'ZipFile');
        });

        describe('xhr', () => {
            let progressCallbackSpy;
            let xhrMock;

            beforeEach(() => {
                progressCallbackSpy = jest.fn();
                xhrMock = {
                    upload: {},
                };
                global.XMLHttpRequest = function XMLHttpRequest() {
                    return xhrMock;
                };
            });

            afterEach(() => {
                global.XMLHttpRequest = undefined;
            });

            it('should pass custom XMLHttpRequest Object with an on progress callback as an option', () => {
                system.uploadApp('ZipFile', progressCallbackSpy);

                expect(typeof apiMock.post.mock.calls[0][2].xhr).toBe('function');
                expect(apiMock.post.mock.calls[0][2].xhr.call()).toBe(xhrMock);
            });

            it('should define the onprogress function onto the upload object of the xhr', () => {
                system.uploadApp('ZipFile', progressCallbackSpy);

                expect(typeof xhrMock.upload.onprogress).toBe('function');
            });

            it('should not call the callback if the progress can not be computed', () => {
                system.uploadApp('ZipFile', progressCallbackSpy);
                xhrMock.upload.onprogress({});

                expect(progressCallbackSpy).not.toBeCalled();
            });

            it('should call the callback spy if the progress can be computed', () => {
                system.uploadApp('ZipFile', progressCallbackSpy);
                xhrMock.upload.onprogress({
                    lengthComputable: true,
                    loaded: 10,
                    total: 50,
                });

                expect(progressCallbackSpy).toBeCalledWith(0.2);
            });
        });
    });

    describe('loadAppStore()', () => {
        // Useful constants!
        const sysVersionMinusTwo = '2.26';
        const sysVersionMinusOne = '2.27';
        const sysVersion = '2.28';
        const sysVersionPlusOne = '2.29';
        const sysVersionPlusTwo = '2.30';

        beforeEach(() => {
            system.setSystemInfo({
                version: sysVersion,
            });

            jest.spyOn(apiMock, 'get')
                .mockReturnValue(Promise.resolve(fixtures.get('/appStore')));
        });

        it('should be a function on the system object', () => {
            expect(typeof system.loadAppStore).toBe('function');
        });

        it('should return a promise', () => {
            expect(system.loadAppStore()).toBeInstanceOf(Promise);
        });

        it('should request the api for the appStore', () => {
            system.loadAppStore();

            expect(apiMock.get).toBeCalledWith('appStore');
        });

        it('should return the compatible apps from the api', () => {
            // Apps 3 and 6 are compatible with 2.28
            const expectedApps = fixtures.get('/appStore').filter((app, i) => [3, 6].includes(i));

            return system.loadAppStore()
                .then((apps) => {
                    expect(apps).toEqual(expectedApps);
                });
        });

        describe('with version 2.24', () => {
            beforeEach(() => {
                system.setSystemInfo({
                    version: '2.24',
                });
            });

            it('should return the compatible apps from the API', () => {
                // Apps 2, 4 and 6 are compatible with 2.24
                const expectedApps = fixtures.get('/appStore').filter((app, i) => [2, 4, 6].includes(i));

                return system.loadAppStore()
                    .then((apps) => {
                        expect(apps).toEqual(expectedApps);
                    });
            });
        });

        describe('with version 2.25', () => {
            beforeEach(() => {
                system.setSystemInfo({
                    version: '2.25',
                });
            });

            it('should return the compatible apps from the API', () => {
                // Apps 1, 2, 4, 5 and 6 are compatible with 2.25
                const expectedApps = fixtures.get('/appStore').filter((app, i) => [1, 2, 4, 5, 6].includes(i));

                return system.loadAppStore()
                    .then((apps) => {
                        expect(apps).toEqual(expectedApps);
                    });
            });
        });

        it('should return the compatible apps', () => {
            // const returnedApps = fixtures.get('/appStore');

            const returnedApps = [
                {
                    versions: [ // One version compatible
                        { min_platform_version: sysVersionMinusTwo, max_platform_version: sysVersionMinusOne },
                        { min_platform_version: sysVersionMinusOne }, // compatible
                    ],
                },
                {
                    versions: [ // Both incompatible
                        { min_platform_version: sysVersionPlusOne, max_platform_version: sysVersionPlusTwo },
                        { min_platform_version: sysVersionPlusTwo },
                    ],
                },
                {
                    versions: [ // Compatible
                        { min_platform_version: sysVersionMinusOne },
                    ],
                },
                {
                    versions: [ // Incompatible
                        { max_platform_version: sysVersionMinusOne },
                    ],
                },
                {
                    versions: [ // One version compatible
                        { minDhisVersion: sysVersionMinusTwo, maxDhisVersion: sysVersionMinusOne },
                        { minDhisVersion: sysVersionMinusOne }, // compatible
                    ],
                },
                {
                    versions: [ // Both incompatible
                        { minDhisVersion: sysVersionPlusOne, maxDhisVersion: sysVersionPlusTwo },
                        { minDhisVersion: sysVersionPlusTwo },
                    ],
                },
                {
                    versions: [ // Compatible
                        { minDhisVersion: sysVersionMinusOne },
                    ],
                },
                {
                    versions: [ // Incompatible
                        { maxDhisVersion: sysVersionMinusOne },
                    ],
                },
            ];

            apiMock.get.mockReturnValue(Promise.resolve(returnedApps));

            return system.loadAppStore()
                .then((apps) => {
                    expect(apps.length).toBe(4);
                });
        });

        it('should return all the apps when compatibility flag is set to false', () => system.loadAppStore(false)
            .then((apps) => {
                expect(apps.length).toBe(fixtures.get('/appStore').length);
            }));

        it('should reject the promise when the request fails', () => {
            apiMock.get.mockReturnValue(Promise.reject('Request for appStore failed'));

            return system.loadAppStore()
                .catch(error => error)
                .then((error) => {
                    expect(error).toBe('Request for appStore failed');
                });
        });

        it('should reject the promise when system.version is not set', () => {
            system.version = undefined;

            return system.loadAppStore()
                .catch(error => error)
                .then((error) => {
                    expect(error.message).toBe('Cannot read property \'major\' of undefined');
                });
        });
    });

    describe('installAppVersion()', () => {
        beforeEach(() => {
            apiMock.post.mockReturnValue(Promise.resolve(''));
        });

        it('should be a function on the system object', () => {
            expect(typeof system.installAppVersion).toBe('function');
        });

        it('should reject the promise when the request fails', () => {
            apiMock.post.mockReturnValue(Promise.reject('Request for installation failed'));

            return system.installAppVersion('PyYnjVl5iGt')
                .catch(error => error)
                .then((errorMessage) => {
                    expect(errorMessage).toBe('Request for installation failed');
                });
        });

        it('should call the api with the correct url', () => system.installAppVersion('PyYnjVl5iGt')
            .then(() => {
                expect(apiMock.post).toBeCalledWith('appStore/PyYnjVl5iGt', '', { dataType: 'text' });
            }));

        it('should resolve the promise without a value', () => system.installAppVersion('PyYnjVl5iGt')
            .then((response) => {
                expect(response).toBeUndefined();
            }));
    });

    describe('uninstallApp()', () => {
        beforeEach(() => {
            apiMock.delete = jest.fn().mockReturnValue(Promise.resolve({}));
        });

        it('should be a function on the system object', () => {
            expect(typeof system.uninstallApp).toBe('function');
        });

        it('should call the api.delete method with the correct url', () => system.uninstallApp('PyYnjVl5iGt')
            .then(() => {
                expect(apiMock.delete).toBeCalledWith('apps/PyYnjVl5iGt');
            }));

        it('should resolve the request even when the api request fails', () => {
            apiMock.delete = jest.fn().mockReturnValue(Promise.reject({}));

            return system.uninstallApp('PyYnjVl5iGt');
        });
    });

    describe('reloadApps()', () => {
        beforeEach(() => {
            jest.spyOn(apiMock, 'update')
                .mockReturnValue(Promise.resolve());

            jest.spyOn(system, 'loadInstalledApps')
                .mockReturnValueOnce(Promise.resolve());
        });

        it('should be a function on the system object', () => {
            expect(typeof system.reloadApps).toBe('function');
        });

        it('should call the update method on the api', () => system.reloadApps()
            .then(() => {
                expect(apiMock.update.mock.calls[0][0]).toBe('apps');
            }));

        it('should call system.loadInstalledApps on success ', () => system.reloadApps()
            .then(() => {
                expect(system.loadInstalledApps).toBeCalled();
            }));

        it('should chain the promise from loadInstalledApps', () => {
            const loadInstalledAppsPromise = Promise.resolve('Apps loaded');

            system.loadInstalledApps.mockReset();
            system.loadInstalledApps.mockReturnValue(loadInstalledAppsPromise);

            return system.reloadApps()
                .then(message => expect(message).toBe('Apps loaded'));
        });

        it('should not call loadInstalledApps when the update request fails', () => {
            apiMock.update.mockReturnValue(Promise.reject());
            jest.spyOn(system, 'loadInstalledApps');

            return system.reloadApps()
                .catch(message => message)
                .then(() => {
                    expect(system.loadInstalledApps).not.toBeCalled();
                });
        });
    });

    describe('compareVersions()', () => {
        let systemVersion;
        let appVersion;

        beforeEach(() => {
            systemVersion = {
                major: 2,
                minor: 23,
                snapshot: true,
            };
            appVersion = {
                major: 2,
                minor: 23,
                snapshot: true,
            };
        });

        it('should be a function on the system class', () => {
            expect(typeof System.compareVersions).toBe('function');
        });

        it('should return 0 for equal versions', () => {
            expect(System.compareVersions(systemVersion, appVersion)).toBe(0);
        });

        it('should return 1 for a larger major system version', () => {
            systemVersion.major = 3;

            expect(System.compareVersions(systemVersion, appVersion)).toBe(1);
        });

        it('should return 1 for a larger minor version', () => {
            systemVersion.minor = 24;

            expect(System.compareVersions(systemVersion, appVersion)).toBe(1);
        });

        it('should return 1 when the app is a snapshot version', () => {
            systemVersion.snapshot = false;
            appVersion.snapshot = true;

            expect(System.compareVersions(systemVersion, appVersion)).toBe(1);
        });

        it('should return -1 when the app is not a snapshot', () => {
            systemVersion.snapshot = true;
            appVersion.snapshot = false;

            expect(System.compareVersions(systemVersion, appVersion)).toBe(-1);
        });

        it('should do correct comparison when a string is passed as a version', () => {
            expect(System.compareVersions('2.15', '2.16')).toBe(-1);
            expect(System.compareVersions('2.20-SNAPSHOT', '2.16')).toBe(4);
        });
    });

    describe('isVersionCompatible()', () => {
        let appVersion;
        let systemVersion;

        beforeEach(() => {
            jest.spyOn(System, 'compareVersions');
            appVersion = {
                min_platform_version: '2.23',
                max_platform_version: '2.23',
            };
            systemVersion = '2.23';
        });

        afterEach(() => {
            System.compareVersions.mockRestore();
        });

        it('should return false when the app is too new', () => {
            expect(System.isVersionCompatible(
                systemVersion,
                Object.assign(appVersion, { min_platform_version: '2.24' }),
            )).toBe(false);
        });

        it('should return false when the app is too old', () => {
            expect(System.isVersionCompatible(
                systemVersion,
                Object.assign(appVersion, { max_platform_version: '2.22' }),
            )).toBe(false);
        });

        it('should return true when the system version is within the app version range', () => {
            appVersion.min_platform_version = '2.20';
            appVersion.max_platform_version = '2.25';

            expect(System.isVersionCompatible(systemVersion, appVersion));
        });

        it('should return true when no version bounds are given', () => {
            appVersion = {};

            expect(System.isVersionCompatible(systemVersion, appVersion)).toBe(true);
        });

        it('should return false when the version is not compatible', () => {
            expect(System.isVersionCompatible('2.22', {
                min_platform_version: '2.17',
                max_platform_version: '2.20',
            })).toBe(false);
        });


        describe('with 2.28 app version format', () => {
            beforeEach(() => {
                appVersion = {
                    minDhisVersion: '2.22',
                    maxDhisVersion: '2.23-SNAPSHOT',
                };
                systemVersion = '2.23';
            });

            it('should return false when the app is too new', () => {
                expect(System.isVersionCompatible(
                    systemVersion,
                    Object.assign(appVersion, { minDhisVersion: '2.24' }),
                )).toBe(false);
            });

            it('should return false when the app is too old', () => {
                expect(System.isVersionCompatible(
                    systemVersion,
                    Object.assign(appVersion, { maxDhisVersion: '2.22' }),
                )).toBe(false);
            });

            it('should return true when the system version is within the app version range', () => {
                appVersion.minDhisVersion = '2.20';
                appVersion.maxDhisVersion = '2.25';

                expect(System.isVersionCompatible(systemVersion, appVersion));
            });

            it('should return true when no version bounds are given', () => {
                appVersion = {};

                expect(System.isVersionCompatible(systemVersion, appVersion)).toBe(true);
            });

            it('should return false when the version is not compatible', () => {
                expect(System.isVersionCompatible('2.22', {
                    minDhisVersion: '2.17',
                    maxDhisVersion: '2.20',
                })).toBe(false);
            });
        });
    });

    describe('getSystem', () => {
        it('should return the same instance on consecutive requests', () => {
            expect(System.getSystem()).toBe(System.getSystem());
        });
    });
});
