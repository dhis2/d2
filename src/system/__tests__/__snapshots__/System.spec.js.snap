// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`System should not be allowed to be called without new 1`] = `"Class constructor System cannot be invoked without 'new'"`;
